### PR TITLE
BUG: Use output when given on numpy.dot C-API branch

### LINKED ISSUE #21081: BUG: scalar.dot(vector) doesn't fill out-array of dtype float16
### Describe the issue:

The output array in the code snippet below should be getting filled with the result of the computation, but it remains unchanged.

### Reproduce the code example:

```python
import numpy as np
x = np.ones((), dtype=np.float16)
y = np.ones((5,), dtype=np.float16)
z = np.zeros((5,), dtype=np.float16)
res = x.dot(y, out=z)
assert np.array_equal(res, y)  # succeeds
assert np.array_equal(z, y)  # fails
```

### Error message:

```shell
Traceback (most recent call last):
  File "/home/mpapadakis/out-half.py", line 7, in <module>
    assert np.array_equal(z, y)  # fails
AssertionError
```

### NumPy/Python version information:

1.22.2 3.10.2 | packaged by conda-forge | (main, Feb  1 2022, 19:29:00) [GCC 9.4.0]

### PR DESCRIPTION
**Pull request by pedro-sl**
Updated the numpy.dot C-API so that it now copies the result to the out parameter and returns it on branch of the function where the correct behaviour was not in place. Added two tests regarding this issue. Closes #21081.

### CODE REVIEW

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 1]
```diff
@@ -1045,6 +1045,20 @@ PyArray_MatrixProduct2(PyObject *op1, PyObject *op2, PyArrayObject* out)
...
+                Py_DECREF(ap1);
+                Py_DECREF(ap2);
+                return NULL;
+            }
+            Py_DECREF(result);
+            Py_DECREF(ap1);
+            Py_DECREF(ap2);
+            Py_INCREF(out);
+            return (PyObject *)out;
+        }
```
Comment (seberg) [inline numpy/core/src/multiarray/multiarraymodule.c:1061]: Thanks!  The file already includes `"number.h"` and that makes calling the ufunc directly easy (and the ufunc supports `out`).

So I think we can simplify this a fair bit by doing something like (i.e. calling the ufunc directly):
```
result = PyObject_CallFunctionObjArgs(n_ops.multiply, ap1, ap2, out, NULL);
/* cleanup */
return result;
```
(that includes removint that weird `result` setup.

Comment (JLameiras) [reply numpy/core/src/multiarray/multiarraymodule.c]: Hi! Thank you for the feedback.
I've simplified the code with your suggestion and everything seems OK.

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -1042,9 +1042,7 @@ PyArray_MatrixProduct2(PyObject *op1, PyObject *op2, PyArrayObject* out)
 #endif
 
     if (PyArray_NDIM(ap1) == 0 || PyArray_NDIM(ap2) == 0) {
-        result = (PyArray_NDIM(ap1) == 0 ? ap1 : ap2);
-        result = (PyArrayObject *)Py_TYPE(result)->tp_as_number->nb_multiply(
-                                        (PyObject *)ap1, (PyObject *)ap2);
+        result = (PyArrayObject *)PyObject_CallFunctionObjArgs(n_ops.multiply, ap1, ap2, out, NULL);
```
Comment (seberg) [inline numpy/core/src/multiarray/multiarraymodule.c:1045]:
```suggestion
        PyObject *mul_res = PyObject_CallFunctionObjArgs(
                n_ops.multiply, ap1, ap2, out, NULL);
```

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 3]
```diff
@@ -1042,9 +1042,7 @@ PyArray_MatrixProduct2(PyObject *op1, PyObject *op2, PyArrayObject* out)
...
         Py_DECREF(ap1);
         Py_DECREF(ap2);
         return (PyObject *)result;
```
Comment (seberg) [inline numpy/core/src/multiarray/multiarraymodule.c:1048]:
```suggestion
        return mul_res;
```

**Review by seberg — APPROVED** [review]
LGTM, thanks.  Also nice that the other test failed now :), could add an explicit test for that, but I don't think it is really necessary so going to do that small edit and then plan to put it in.

**Pull request merged**